simtfl/bft/__init__.py

```diff
@@ -0,0 +1,142 @@
+    `ceiling(n * 2/3)`.
+    """
+    return (n * 2 + 2) // 3
+
+
+class PermissionedBFTBase:
+    """
+    This class is used for the genesis block in a permissioned BFT protocol
+    (which is taken to be notarized, and therefore valid, by definition).
+
+    It is also used as a base class for other BFT block and proposal classes.
+    """
+    def __init__(self, n, t):
+        """
+        Constructs a genesis block for a permissioned BFT protocol with
+        `n` nodes, of which at least `t` must sign each proposal.
+        """
+        self.n = n
+        self.t = t
+        self.parent = None
+
+
+class PermissionedBFTBlock(PermissionedBFTBase):
+    """
+    A block for a BFT protocol. Each non-genesis block is based on a
+    notarized proposal.
+
+    BFT blocks are taken to be notarized, and therefore valid, by definition.
+    """
+
+    def __init__(self, proposal):
+        super().__init__(proposal.n, proposal.t)
+
+        assert proposal.is_notarized()
+        self.proposal = proposal
+
+
+class PermissionedBFTProposal(PermissionedBFTBase):
+    """A proposal for a BFT protocol."""
+
+    def __init__(self, parent):
+        """
+        Constructs a `PermissionedBFTProposal` with the given parent
+        `PermissionedBFTBlock`. The parameters are determined by the parent
+        block.
+        """
+        super().__init__(parent.n, parent.t)
+        self.parent = parent
+        self.signers = set()
+
+    def assert_valid(self):
+        """
+        Assert that this proposal is valid. This does not assert that it is
+        notarized. This should be overridden by subclasses.
+        """
+        pass
+
+    def is_valid(self):
+        """Is this proposal valid?"""
+        try:
+            self.assert_valid()
+            return True
+        except AssertionError:
+            return False
+
+    def assert_notarized(self):
+        """
+        Assert that this proposal is notarized. A `PermissionedBFTProposal`
+        is notarized iff it is valid and has at least the threshold number of
+        signatures.
+        """
+        assert self.is_valid()
+        assert len(self.signers) >= self.t
+
+    def is_notarized(self):
+        """Is this proposal notarized?"""
+        try:
+            self.assert_notarized()
+            return True
+        except AssertionError:
+            return False
+
+    def add_signature(self, index):
+        """
+        Record that the node with the given `index` has signed this proposal.
+        If the same node signs more than once, the subsequent signatures are
+        ignored.
+        """
+        self.signers.add(index)
+
+
+import unittest
+
+
+class TestPermissionedBFT(unittest.TestCase):
+    def test_basic(self):
+        # Construct the genesis block.
+        current = PermissionedBFTBase(5, 2)
+
+        for i in range(2):
+            proposal = PermissionedBFTProposal(current)
+            self.assertTrue(proposal.is_valid())
```

Comment:
```suggestion
            proposal.assert_valid()
            self.assertTrue(proposal.is_valid())
```

Reply:
Why are both lines necessary? Is this "verification-in-depth" which redundantly checks the same condition through different code paths, or are those checks intentionally distinct?

Reply:
We call `proposal.assert_valid()` first because if the proposal is (incorrectly) not valid, that will tell us why — whereas `is_valid()` will catch the exception and lose error information. Then we test that `is_valid()` is implemented correctly for the True case.

Reply:
I still don't understand the second statement.

Aren't the two lines entirely equivalent, so long as `assert_valid` is defined as follows?

```
def assert_valid(self):
  assert self.is_valid()
```

I just looked it up, and it seems equivalent, though the dependency is inverted so we have:

```
    def is_valid(self) -> bool:
        """Is this proposal valid?"""
        try:
            self.assert_valid()
            return True
        except AssertionError:
            return False
```

My question is "is there ever a case where `assert_valid` and `is_valid` will not be entirely equivalent modulo boolean return vs raising AssertionError?

That _could_ be the case if a subclass of `PermissionedBFTProposal` overrides both methods and makes them logically non-equivalent. So I would refine my question to these two clauses:

- Is it possible / useful for subclasses to override these two methods in a manner in which they aren't logically equivalent (modulo raise-vs-return)?
- If not, is the purpose of these two lines to ensure all subclasses have implemented those two methods as logically equivalent (module raise-vs-return)?

If the answer to the first clause is 'no', then we could add a doc string or maybe a type signature to `def is_valid` to prohibit overriding that method as a maintenance improvement / clarification.